src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs

```diff
@@ -82,35 +81,29 @@ public sealed override async Task WriteResponseBodyAsync(OutputFormatterWriteCon
                 // the behavior you get when the user does not declare the return type and with Json.Net at least at the top level.
                 var objectType = context.Object?.GetType() ?? context.ObjectType ?? typeof(object);
                 await JsonSerializer.SerializeAsync(writeStream, context.Object, objectType, SerializerOptions);
-
-                // The transcoding streams use Encoders and Decoders that have internal buffers. We need to flush these
-                // when there is no more data to be written. Stream.FlushAsync isn't suitable since it's
-                // acceptable to Flush a Stream (multiple times) prior to completion.
-                if (writeStream is TranscodingWriteStream transcodingStream)
-                {
-                    await transcodingStream.FinalWriteAsync(CancellationToken.None);
```

Comment:
@GrabYourPitchforks looks like this is rolled in to Dispose \ DisposeAsync. The new behavior would mean we might attempt to write to the response after an exception has been thrown part way during serialization as part of the finally block. 

Is there a different way to approach this?

Reply:
Can you clarify your concern? If there's an exception while trying to process the output stream then presumably the normal exception handling logic that you have wrapped around this method should kick in.

Reply:
There's a small edge case - a write in the finally block might result in a different exception suppressing the originally caught exception. There isn't an obvious code path where this is problematic, but it was something we tried to avoid with the previous design.

Reply:
👀

Reply:
does the call to `FlushAsync` address this?

Reply:
Generally speaking, whenever you have a `Dispose` or `DisposeAsync` method which might throw an exception, you can use a pattern similar to below to swallow any inner exception and to let the outer exception propagate cleanly.

```cs
ExceptionDispatchInfo? edi = null;
try
{
    DoMainWork();
}
catch (Exception ex)
{
    edi = ExceptionDispatchInfo.Capture(ex);
}
finally
{
    try
    {
        DoDispose();
    }
    finally
    {
        edi?.Throw();
    }
}
```

Reply:
I've updated the code to use this pattern.